extension/src/popups/MinervaMenu.js

```diff
@@ -21,21 +23,27 @@ function displayClassRankings() {
     });
 }
 
-window.onload = function() {
-    chrome.storage.local.get(["student_name"], (result) => {
-        if (result.hasOwnProperty("student_name")) {
-            transitionToMainMenu(result["student_name"]);
+function diplayCurrentStudentInfo(studentName) {
```

Comment:
```
async ..... () {
    let result = await new Promise((resolve, reject) {
        chrome.runtime.sendMessage({
            type: "FetchCurrentStudentInfo",
            studentName: name
        }, (response) => {
            resolve(response);
        });
   });
    if (result.status) {
        renderMoreStuff
    }
}
```

Reply:
Is this a suggestion or are you telling me to implement this way.
Because I feel like it's more readable the way I implemented.
There's a clear view of control flow through the process; it's clear to a reader that the we check the response and execute different code based off that response.

Reply:
I think you should adopt `async/await` in this file. `Async/await` more or less is an upgrade to the old Javascript callback paradigm.

Looking at the function alone, the logic is fine. However, let's look at how this function is used in the `onload` handler. Specifically, this function is called regardless whether an user is logged in or not (this is a red flag already, we are not caching/saving anything since it is called on every popup onclick).

To mitigate the issue above, you can certainly change this call to be within the resolve branch (specifically, line `44`). Now, if we trace the callchain, we will have something like this

```
fetchName -> callback -> fetchProfile -> callback -> renderProfile
                                  (in parallel)    -> fetchRanking -> callback -> renderRanking
```
The logic above is perfect yet we have this long and nested callchain. From the top level (`fetchName`), we are not going to know the end

`Async/Await` is introducted to simplify this process/callchain

What I imagine with #97 is that you can have the following,

```
function getRankingPromise() {
    return new Promise ( .....
         runtime.message
             type: GetRanking
             resolve()
         reject()
}

function getProfilePromise() {
    return new Promise ( .....
         runtime.message
             type: GetProfile
             resolve()
         reject()
}


function getNamePromise() {
    return new Promise ( .....
         runtime.message
             type: GetName
             resolve()
         reject()
}

async function entry() {
     // called after checking that an user is logged in, see #97
     let nameRes = await getNamePromise()
     let profileRes = await getProfilePromise()
     let rankingRes = await getRankingPromise()

     // Each function will check the status field
     renderName(nameRes)
     renderProfile(profileRes)
     renderRanking(rankingRes)
}
```

As shown above, we are able to glue all logic together have them all in the same function. Now, if you worry about parallelization, you can look up `Promise.all`, I would say we are ok without it.

Reply:
I agree but I wish I didn't.
I'll refactor.

Reply:
Add-on. I agree on this paradigm for renderName/getName.
But not for renderProfile and renderRanking. My reasoning is because getName is returning a value we actually need in our execution logic. RenderProfile is dependent on getName. But after we have our studentName, renderProfile can execute on it's own and doesn't need to follow this promise paradigm because nothing else is dependent on it.

RenderRanking is the same way but it doesn't need any input variables. So I don't think renderRanking needs to follow this paradigm.

Reply:
> RenderProfile is dependent on getName. But after we have our studentName, renderProfile can execute on it's own and doesn't need to follow this promise paradigm because nothing else is dependent on it.

If I understand this argument correctly, you are saying that `renderProfile` is dependent on `getProfile` which is dependent on `getName`. That is a good point.  You are correct that the paradigm attached above is simplifying everything into a single chain. There are several independent dependency chains in the function, by adopting the new baseline paradigm is eliminating the underlying dependencies between function (e.g. what do we need before we render).  The biggest benefit I see with the new approach is that we have a holistic view of  what we fetch and what we render in one place. The problem at the end is essentially a trade-off between readability and concurrency. You can look into promise APIs to see how to can squeeze out concurrency while maintaining a high level of readability

Reply:
Refactored. Take a look.